hilti/toolchain/src/compiler/visitors/normalizer.cc

```diff
@@ -0,0 +1,218 @@
+            }
+
+            Expression index_assign =
+                hilti::expression::UnresolvedOperator(hilti::operator_::Kind::IndexAssign,
+                                                      {std::move(map), std::move(key), std::move(value)},
+                                                      assign.meta());
+
+            logChange(p.node, index_assign);
+            p.node = std::move(index_assign);
+            modified = true;
+            return;
+        }
+
+        // Rewrite assignments involving struct elements to use the non-const member operator.
+        if ( auto member_const = lhs.tryAs<operator_::struct_::MemberConst>() ) {
+            auto new_lhs = operator_::struct_::MemberNonConst::Operator().instantiate(member_const->operands().copy(),
+                                                                                      member_const->meta());
+            Expression n = expression::Assign(new_lhs, assign.source(), assign.meta());
+            logChange(p.node, n);
+            p.node = std::move(n);
+            modified = true;
+            return;
+        }
+
+        // Rewrite assignments involving tuple ctors on the LHS to use the
+        // tuple's custom by-element assign operator. We need this to get
+        // constness right.
+        auto lhs_ctor = lhs.tryAs<expression::Ctor>();
+        if ( lhs_ctor && lhs_ctor->ctor().isA<ctor::Tuple>() ) {
+            if ( expression::isResolved(assign.source()) && expression::isResolved(assign.target()) ) {
+                auto n = operator_::tuple::CustomAssign::Operator().instantiate({assign.target(), assign.source()},
+                                                                                assign.meta());
+                logChange(p.node, n);
+                p.node = std::move(n);
+                modified = true;
+                return;
+            }
+        }
+    }
+
+    void operator()(const statement::If& n, position_t p) {
+        if ( n.init() && ! n.condition() ) {
+            auto cond = expression::UnresolvedID(n.init()->id());
+            logChange(p.node, cond);
+            p.node.as<statement::If>().setCondition(std::move(cond));
+            modified = true;
+        }
+    }
+
+    void operator()(const statement::Switch& s, position_t p) { p.node.as<statement::Switch>().preprocessCases(); }
+
+    void operator()(const type::Library& t, position_t p) {
+        auto& type = p.node.as<Type>();
+
+        if ( ! type.cxxID() )
+            // Make it equal to types with the same C++ representation.
+            type.setCxxID(ID(t.cxxName()));
+    }
+
+    void operator()(const type::Struct& t, position_t p) { type::Struct::setSelf(&p.node); }
+};
+
+} // anonymous namespace
+
+struct VisitorClearCanonicalIDs : public visitor::PreOrder<void, VisitorClearCanonicalIDs> {
+    result_t operator()(const Declaration& d, position_t p) { p.node.as<Declaration>().setCanonicalID(ID()); };
```

Comment:
I see inconsistent IDs for hooks implemented in another module.

Example:

```
$ cat a.spicy
module a;

public type X = unit {
    x: uint8;
};
```

```
$ cat b.spicy
module b;

import a;

on a::X::x { print "from b", $$; }
```

```
$ spicyc a.spicy b.spicy  -D ast-final  -p -o /dev/null 2>&1 -g | grep b::a::X::
[debug/ast-final]     - spicy::declaration::UnitHook (b.spicy:5:1) [canon-id: b::a::X::x] [@s:7f82b3706080]
[debug/ast-final]         | __dd -> declaration::Expression %163 <linkage="private"> [canon-id: b::a::X::x::__dd] [@d:7f82a3535bd0] (type: uint<8> @t:7f82b37042f0)
[debug/ast-final]           | __dd -> declaration::Expression %163 <linkage="private"> [canon-id: b::a::X::x::__dd] [@d:7f82a3535bd0] (type: uint<8> @t:7f82b37042f0)
[debug/ast-final]         - declaration::Expression %163 <linkage="private"> [canon-id: b::a::X::x::__dd] [@d:7f82a3535bd0]
[debug/ast-final]       | a::X::__on_x -> declaration::Function %305 <linkage="struct" parent_type="%1"> [canon-id: b::a::X::__on_x] [@d:7f82a34250c0] ([@d:7f82a34250c0])
[debug/ast-final]     - declaration::Function %305 <linkage="struct" parent_type="%1"> (b.spicy:5:1) [canon-id: b::a::X::__on_x] [@d:7f82a34250c0]
[debug/ast-final]         | __dd -> declaration::Parameter %306 <is_type_param="false" kind="in"> [canon-id: b::a::X::__on_x::__dd] [@d:7f82a3421670] ([@d:7f82a3421670])
[debug/ast-final]           - declaration::Parameter %306 <is_type_param="false" kind="in"> (b.spicy:5:1) [canon-id: b::a::X::__on_x::__dd] [@d:7f82a3421670]
```

Here the symbols in `a` as seen from `b` sometimes are prefixed with `b::`.

For `$$` this seems to be clearly incorrect as `$$` in either `a` or `b` should always have the same identity.

I am not so sure about the name of the hook: for the optimizer I would like to have a simple way to detect whether that function impl corresponds to some decl for which the canon id seem nice, but currently does not work. OTOH one could also argue that the decl and the impl are actually different entities and one would need to do extra work to connect them.

Reply:
Ok, fix coming.

Reply:
Still seeing some incorrect canon IDs, e.g., example reduced from `tests/hilti/types/struct/params.hlt`:

```
$ cat x.hlt
module Foo {

type S = struct {
    method void test(string A, real B);
};

method void S::test(string A, real B) {}

global auto s1 = new S();
s1.test("a", 3);

}
```

```
$ hiltic -g -p -o /dev/null -D ast-final x.hlt 2>&1 | grep '::test]'
[debug/ast-final]       | S::test -> declaration::Function %2 <linkage="struct" parent_type="%1"> [canon-id: Foo::S::test] [@d:7f8f00c12360] ([@d:7f8f00c12360])
[debug/ast-final]               - declaration::Field <cc="<standard>"> (x.hlt:4:5) [canon-id: Foo::test] [@d:7f8f00c117e0]
[debug/ast-final]         - declaration::Field <cc="<standard>"> (x.hlt:4:5) [canon-id: Foo::test] [@d:7f8f00c117e0]
[debug/ast-final]     - declaration::Function %2 <linkage="struct" parent_type="%1"> (x.hlt:7:1) [canon-id: Foo::S::test] [@d:7f8f00c12360]
```

I see here both of `Foo::S::test` and `Foo::test` while there probably should only be the former. Curiously, when removing parameters from the function or just removing the implicit conversion this would have worked.

Reply:
Ah, that's a regression, I took out a check to prevent this while I was trying to make sure all canonical IDs are set. However, turns out that's not a feasible objective: we have a legitimate case without a canonical IDs, so I just reverted that: https://github.com/zeek/spicy/pull/984/commits/30395bc979b90002da4c0af0724ea5d5c03149dc

Reply:
Seems finally fixed now.